byron/ledger/executable-spec/src/Ledger/UTxO/Generators.hs

```diff
@@ -0,0 +1,273 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Ledger.UTxO.Generators where
+
+import Control.Applicative (empty)
+import Data.Bitraversable (bitraverse)
+import qualified Data.Map.Strict as M
+
+import Ledger.Core hiding (Range, range)
+import Ledger.UTxO hiding (utxo)
+
+import Hedgehog (Gen, Property, Range, (===), assert, forAll, property)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import Hedgehog.Internal.Gen
+  (atLeast, ensure, mapGenT, runDiscardEffect, toTree, toTreeMaybeT)
+import Hedgehog.Internal.Tree (NodeT(..), TreeT(..), treeValue)
+import qualified Hedgehog.Internal.Tree as Tree
+
+
+--------------------------------------------------------------------------------
+-- Initial TxOut generator
+--------------------------------------------------------------------------------
+
+-- | Generate a set of initial @TxOut@s from a set of @Addr@s
+genInitialTxOuts :: [Addr] -> Gen [TxOut]
+genInitialTxOuts = Gen.filter (not . null)
+  . genTraverseSubsequence (\a -> TxOut a <$> genLovelace)
+
+genLovelace :: Gen Lovelace
+genLovelace = Lovelace . fromIntegral <$> Gen.word32 (Range.linear 1 10000)
+
+-- | Generate a subsequence of a list of values and traverse the subsequence
+--   with a generator producer
+genTraverseSubsequence :: (a -> Gen b) -> [a] -> Gen [b]
+genTraverseSubsequence genA as =
+  mapGenT (TreeT . (interleaveTreeT . nodeValue =<<) . runTreeT) $ do
+    sub <- Gen.subsequence as
+    traverse (toTreeMaybeT . genA) sub
+
+-- | Generate a list using @genTraverseSubsequence@
+genList :: Range Int -> Gen a -> Gen [a]
+genList range gen = Gen.sized $ \size ->
+  ensure (atLeast $ Range.lowerBound size range) $ genTraverseSubsequence
+    (const gen)
+    (replicate (Range.upperBound size range) ())
+
+-- | Temporarily defined here until hedgehog exposes this function
+interleaveTreeT :: Monad m => [TreeT m a] -> m (NodeT m [a])
+interleaveTreeT = fmap Tree.interleave . traverse runTreeT
+
+
+--------------------------------------------------------------------------------
+-- Tx generator
+--------------------------------------------------------------------------------
+
+-- | Generate a valid transaction from a given @UTxO@
```

Comment:
BTW, don't we need to use `'` instead of `@` so that haddock will create references?

Reply:
Sure, I always forget which one is the linking one

Reply:
I just double checked just in case https://www.haskell.org/haddock/doc/html/ch03s08.html

FWIW I would have chosen the opposite :)